### ISSUE TITLE
Compatibility renderer: lights seem brighter when their shadows are enabled

### ISSUE THREAD

**Issue by luciusponto** [godotengine/godot#90259]
### Tested versions

- Reproducible in: 4.2.0, 4.3.dev5

### System information

Godot v4.3.dev5 - Windows 10.0.19045 - GLES3 (Compatibility) - NVIDIA GeForce GTX 1060 6GB (NVIDIA; 31.0.15.3713) - Intel(R) Core(TM) i5-3470 CPU @ 3.20GHz (4 Threads)

### Issue description

When using the compatibility renderer, enabling shadows on a light will make lit surfaces brighter. This happens for directional (over the whole scene), omni and spotlights.

If the ambient light is made black, the problem goes away. If shadows are disabled for all lights, the problem also goes away. If multiple lights have shadows turned on, the problem is more pronounced than if a single light has shadows turned on. It looks like the ambient light might be applied one extra time for each light with shadows enabled.

This seems to have been introduced when compatibility renderer shadows were first added. Forward plus and mobile renderers are not affected.

Comparison of a scene with a default world environment lit by a single directional light with shadows disabled and enabled:

![Compatibility_ShadowsOff](https://github.com/godotengine/godot/assets/23417591/4cee3255-718c-48da-b3d5-be1b1e9eb2b7)
![Compatibility_ShadowsOn](https://github.com/godotengine/godot/assets/23417591/bb396fb0-1e1e-4a88-9bb0-d36a4f778c72)

### Steps to reproduce

Run the project with F5. Press 'S' to toggle shadows on the directional light on and off. Observe how surfaces get brighter when shadows are on.

### Minimal reproduction project (MRP)

[CompatShadowRepro.zip](https://github.com/godotengine/godot/files/14884361/CompatShadowRepro.zip)

**Comment by Arthas92t** [comment]
I think issue come from the way convert linear_to_srgb, in scene.glsl with 1 light which have shadow, for now:
frag_color.rgb = linear_to_srgb(emission + ambient_light) +  linear_to_srgb(additive_light_color);
while i think it should be 
frag_color.rgb = linear_to_srgb(emission + ambient_light + additive_light_color);

So if i apply this path which remove 2 time convert linear_to_srgb and only convert for once after all calculate, color will look like vulkan in case we use 1 shadow.
![image](https://github.com/godotengine/godot/assets/1473739/d6741524-9d13-4fa0-b5e1-594a28708a91)

But in case we have more than 1 shadow, engine will run into multiple light pass and blend function still have problem when we try to use srgb.

**Comment by DarioDaF** [comment]
Running the editor itself as `--rendering-method gl_compatibility --rendering-driver opengl3` doesn't seem to actually render accordingly the scene anyway, it's ok not to have compatibility or behaviour fix, but there must be a way to actually see in editor the same as the exported scene (current release and 4.3 dev5 show brighter in editor and duller in run)

**Comment by MichaelDawe** [comment]
Can confirm this is still an issue in Godot v4.3.beta1.  It is visible in the editor window, with the default directional light and world environment.  When turning off the world environment it looks very similar to the other renderers, however I tested deleting my environment in a larger project and running it and the issue is still present, so its not solely related to the environment.  Don't know much about this stuff but I'm wondering if some of the default shaders were ported from the Vulcan renderers and the values are coming out differently?

**Comment by Calinou** [comment]
This makes me wonder if we should add a light energy multiplier that is *only* used when using the Compatibility rendering method. This would make setting up fallbacks a lot easier, without needing to attach a script to each light that reads node metadata (or something along those lines).

This could also be worth considering for lights that have a projector texture, as you generally want those to have lower energy when the project texture can't be used (otherwise, the light ends up too bright). I believe UE5 has a similar property for projector fallbacks in its lighting system.

**Comment by MichaelDawe** [comment]
This issue wasn't present in 4.2.2.stable.  I know that the compatibility renderer wasn't feature complete in that version, but I feel that this is a regression that is (potentially) a simple mistake along the way, as lights and shadows and world environment all worked, despite certain features like fog not working.

**Comment by mcjohnalds** [comment]
The 0.25 trick looks pretty good to me. I'm content with having a script on every DirectionalLight3D.

**Comment by outfrost** [comment]
I have a Godot 3 project which I'm trying to port to Godot 4.

In Godot 4 with GLES3, I observe the behaviour described in this issue (though I haven't yet determined exactly how much brighter the lights are with shadows on).

With shadows on, the lighting looks much closer to how it does in Godot 3.

In Godot 3 with GLES3, the lights' effective brightness doesn't change if I toggle their shadows.

This leads me to believe that, while the light intensity values might inherently work differently between Forward+ and Compatibility, the inconsistency on Compatibility between shadows off vs on is a bug that may need to be addressed separately.

**Comment by MichaelDawe** [comment]
Hello all.  I've been looking into this again after 4.4 which enables automatic roll-back to the compatibility renderer if Vulkan is unsupported.  I'm pretty sure I've determined the issue.  I will be attempting to fix it, if I succeed I will create a pull request.  
The issue appears to be that ambient light is added over again for every light with shadows enabled, not to do with colour space conversions.

**Comment by DarioDaF** [comment]
I can confirm that the effect I have seen seems dependent also on number of lights so @MichaelDawe I think you are on a good track in my opinion.
Thank you for pinning it down hope you get to find the source of the problem

**Comment by MichaelDawe** [comment]
I have determined the cause of the issue!  It is caused by **ambient light**, **emissive light**, or other **lights that don't cast shadows**.

Both **ambient light** and **emissive light** are added in the **_first pass_**, along with **lights that don't cast shadows**.
The **_second pass_** (used for lights with shadows) multiplies the lighting with "**albedo**" which is the result of the **_first pass_**, not the actual **albedo** of the materiel, (i.e. texture * vertex colours * albedo(the setting in material options), etc.).

It's getting late here, and the solution is looking fairly complicated - I'm not very familiar with shaders or C++, I'll keep digging into it in my free time but maybe this can help someone who knows more about it find a solution.

My current proposal is to calculate only lighting in the first pass, add the results of the second pass to it, and multiply the result of both passes with albedo at the end of the second pass.  This would require a uniform/define (e.g. MATERIAL_HAS_SECOND_PASS) that tells the first pass whether or not a second pass is going to be applied (it is only applied if the object is in range of a light which casts shadows).

In my opinion this is clearly a bug, and with the recent change to automatically fall back to the compatibility renderer on systems which don't support Vulkan (a change I greatly appreciate btw) I think this is more of a priority.

**Comment by MichaelDawe** [comment]
Would anyone be able to point me towards the C++ code which set the uniforms etc. for the shader?  Thanks :)

**Comment by DarioDaF** [comment]
@MichaelDawe https://github1s.com/godotengine/godot/blob/master/drivers/gles3/rasterizer_scene_gles3.cpp#L1295-L1338
Here there is some code handling data for lights (it's the light data tho not the uniforms themselves)

**Comment by MichaelDawe** [comment]
Yes, my initial conclusion was incorrect, after outputting just albedo from the additive lighting passes I was able to see it looked as expected.  

I'll look at the shaders again tomorrow. 

The effects of overlapping with ambient, emissive, or shadowless lights is confirmed via the following tests which can be performed in-editor, as well as some tests with changing the shader outputs.

Test for ambient (on a non-emissive material with no shadowless lights affecting it):

1) With ambient disabled or set to black, shadowed light looks correct i.e. the same as shadowless and the same as other rendering backends.  Importantly, adding more shadowed lights overlapping each other is still correct, this leads me to think that colour space conversions are not the issue as they are applied to the shadowed lights between passes, however I'm not certain.

2) With ambient set to a solid colour the shadowed lights become overly bright in just the ambient channel (most visible when it is somewhat dim (~0.3 is good) because if it's 1 the output is clipped and the effect becomes undetectable).  This can be seen by eye but also confirmed by post-processing that removes the ambient channel altogether, in which case the result is the same whether shadows are enabled or not.

Test for emissive:
The same as for ambient, with ambient disabled and emissive enabled.  Results the same as ambient.

Test for overlapping shadowless lights:
Same again.

It's possible that other things can affect it too, such as rim lighting, though I haven't got any conclusive evidence yet.  It also appears that specular is working correctly though again I'm not certain.

**Comment by clayjohn** [comment]
@MichaelDawe don't forget that the total brightness impacts how "off" things look. Two bright lights will blend poorly compared to two dim lights. Similarly two dim lights will blend poorly when there is ambient lighting compared to if there is no ambient lighting. 

You should test again with ambient light disabled but with two brighter shadowed lights.

**Comment by MichaelDawe** [comment]
> [@MichaelDawe](https://github.com/MichaelDawe) don't forget that the total brightness impacts how "off" things look. Two bright lights will blend poorly compared to two dim lights. Similarly two dim lights will blend poorly when there is ambient lighting compared to if there is no ambient lighting.
> 
> You should test again with ambient light disabled but with two brighter shadowed lights.

You're right, thanks :)  I wonder if it is an issue with clipped values.  I can't remember where I read this but I believe the compatibility renderer doesn't support HDR while the others do.  I think that might be an OpenGL limitation, in which case it'll be harder to fix.

**Comment by martin-lorentzon** [comment]
Here is a very jarring example of what I believe might be the same issue described in this post.
I initially thought this was a problem related to tone mapping. Very confusing.
  
![alt text](https://i.gyazo.com/435dfaf335710096b8e5d76301314a0e.gif)

**Comment by ghost** [comment]
Lights seem to combine with Ambient Light, you can see it when shadows are disabled and you are very slowly increasing ambient light, as soon as you turn on shadows they stop this behavior and stay locked at their base value as if ambient light is set to 0.

Ambient Energy 0
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/49262091-70f0-4251-abdc-7558da46fa47" />

Ambient Energy 16
Edit: If shadow is enabled the middle of the light is combining but the outer is not. If shadow is off only the outer is combining.
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/62458d69-6a60-4f6e-9201-d5b88734c12f" />

Edit: Without DirectionalLight3D Shadow
Ambient Energy 0
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/e1a2affd-fc42-49a5-bb89-3d7492ff5028" />

Ambient Energy 16
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/d496a718-103a-40fc-84d2-c8652642ae46" />

And Without DirectionalLight3D itself
Ambient Energy 0
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/1fb90258-a667-4228-8769-9dde42bd3142" />

Ambient Energy 16
<img width="1920" height="1080" alt="Image" src="https://github.com/user-attachments/assets/96458c6a-c11e-4841-8f69-eb53d0e3b8bc" />
